fdbclient/MonitorLeader.actor.cpp

```diff
@@ -98,39 +98,56 @@ std::string ClusterConnectionString::getErrorString(std::string const& source, E
 }
 
 ACTOR Future<Void> resolveHostnamesImpl(ClusterConnectionString* self) {
-	std::vector<Future<Void>> fs;
-	for (auto const& hostName : self->hostnames) {
-		fs.push_back(map(INetworkConnections::net()->resolveTCPEndpoint(hostName.host, hostName.service),
-		                 [=](std::vector<NetworkAddress> const& addresses) -> Void {
-			                 NetworkAddress addr = addresses[deterministicRandom()->randomInt(0, addresses.size())];
-			                 addr.flags = 0; // Reset the parsed address to public
-			                 addr.fromHostname = NetworkAddressFromHostname::True;
-			                 if (hostName.isTLS) {
-				                 addr.flags |= NetworkAddress::FLAG_TLS;
-			                 }
-			                 self->addResolved(hostName, addr);
-			                 return Void();
-		                 }));
-	}
-	wait(waitForAll(fs));
-	std::sort(self->coords.begin(), self->coords.end());
-	if (std::unique(self->coords.begin(), self->coords.end()) != self->coords.end()) {
-		throw connection_string_invalid();
+	loop {
+		if (self->status == UNRESOLVED) {
+			self->status = RESOLVING;
+			std::vector<Future<Void>> fs;
+			for (auto const& hostname : self->hostnames) {
+				fs.push_back(map(INetworkConnections::net()->resolveTCPEndpoint(hostname.host, hostname.service),
+				                 [=](std::vector<NetworkAddress> const& addresses) -> Void {
+					                 NetworkAddress address =
+					                     addresses[deterministicRandom()->randomInt(0, addresses.size())];
+					                 address.flags = 0; // Reset the parsed address to public
+					                 address.fromHostname = NetworkAddressFromHostname::True;
+					                 if (hostname.isTLS) {
+						                 address.flags |= NetworkAddress::FLAG_TLS;
+					                 }
+					                 self->addResolved(hostname, address);
+					                 return Void();
+				                 }));
+			}
+			wait(waitForAll(fs));
+			std::sort(self->coords.begin(), self->coords.end());
+			if (std::unique(self->coords.begin(), self->coords.end()) != self->coords.end()) {
+				self->status = UNRESOLVED;
+				self->resolveFinish.trigger();
+				throw connection_string_invalid();
+			}
+			self->status = RESOLVED;
+			self->resolveFinish.trigger();
+			break;
+		} else if (self->status == RESOLVING) {
+			wait(self->resolveFinish.onTrigger());
+			if (self->status == RESOLVED) {
+				break;
+			}
+			// Otherwise, this means other threads failed on resolve, so here we go back to the loop and try to resolve
+			// again.
+		} else {
+			// status is RESOLVED, nothing to do.
+			break;
+		}
 	}
-	self->hasUnresolvedHostnames = false;
 	return Void();
 }
 
 Future<Void> ClusterConnectionString::resolveHostnames() {
-	if (!hasUnresolvedHostnames) {
-		return Void();
-	} else {
-		return resolveHostnamesImpl(this);
-	}
+	return resolveHostnamesImpl(this);
 }
 
 void ClusterConnectionString::resolveHostnamesBlocking() {
```

Comment:
This actually means a synchronous call rather than an async call. Do you want me to rename it to resolveHostnamesSync()?

Reply:
Call it `resolveHostnames`, but seems that name is already used. Maybe `resolveAllHostnames`?

Reply:
`resolveAllHostnames` doesn't sound good to me. Another option is name this `resolveHostnames`, and name the other one `resolveHostnamesAsync`. But it's also a little weird because `resolveHostnamesAsync` should be the default one to use. Only when we cannot call it (in non-actor functions) should we call the synchronous one.